contracts/SP1_SpokePool.sol

```diff
@@ -0,0 +1,167 @@
+
+    error NotHubPoolStore();
+    error NotTarget();
+    error AdminCallAlreadySet();
+    error SlotValueMismatch();
+    error AdminCallNotValidated();
+    error DelegateCallFailed();
+    error AlreadyReceived();
+    error NotImplemented();
+
+    // All calls that have admin privileges must be fired from within the receiveL1State method that validates that
+    // the input data was published on L1 by the HubPool. This input data is then executed on this contract.
+    // This modifier sets the adminCallValidated variable so this condition can be checked in _requireAdminSender().
+    modifier validateInternalCalls() {
+        // Make sure adminCallValidated is set to True only once at beginning of the function, which prevents
+        // the function from being re-entered.
+        if (_adminCallValidated) {
+            revert AdminCallAlreadySet();
+        }
+
+        // This sets a variable indicating that we're now inside a validated call.
+        // Note: this is used by other methods to ensure that this call has been validated by this method and is not
+        // spoofed.
+        _adminCallValidated = true;
+
+        _;
+
+        // Reset adminCallValidated to false to disallow admin calls after this method exits.
+        _adminCallValidated = false;
+    }
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor(
+        address _verifier,
+        address _helios,
+        bytes32 _acrossCallProgramVKey,
+        address _hubPoolStore,
+        address _wrappedNativeTokenAddress,
+        uint32 _depositQuoteTimeBuffer,
+        uint32 _fillDeadlineBuffer
+    ) SpokePool(_wrappedNativeTokenAddress, _depositQuoteTimeBuffer, _fillDeadlineBuffer, OFT_FEE_CAP) {
+        verifier = _verifier;
+        helios = _helios;
+        acrossCallProgramVKey = _acrossCallProgramVKey;
+        hubPoolStore = _hubPoolStore;
+    }
+
+    function initialize(
+        uint32 _initialDepositId,
+        address _crossDomainAdmin,
+        address _withdrawalRecipient
+    ) public initializer {
+        __SpokePool_init(_initialDepositId, _crossDomainAdmin, _withdrawalRecipient);
+    }
+
+    /**
+     * @notice This can be called by an EOA to relay call data stored on the HubPool on L1 into this contract.
+     * @dev Consider making this an onlyOwner function so that only a privileged EOA can relay state. This EOA
+     * wouldn't be able to tamper with the _publicValues but we can reduce the chance of replay-attacks this way if
+     * we set the EOA to a trusted actor.
+     * @param _publicValues L1 contract state we want to relay into this contract. Contains a message that will
+     * be treated as calldata for a delegatecall into this contract.
+     * @param _proofBytes Proof bytes for the public values.
+     * @param _head Head number related to _publicValues.
+     */
+    function receiveL1State(
+        bytes calldata _publicValues,
+        bytes calldata _proofBytes,
```

Comment:
Because this function _verifies the proof_ (rather than asking for the slot from the verifier contract), we need to be sure the helios contract allows the same proof to be replayed. For instance, we wouldn't want an attacker to be able to frontrun the proof on the verifier contract and cause our verification in this function to fail (DoS).

Reply:
Is there anything to be done in this contract to support this use case? 

> we need to be sure the helios contract allows the same proof to be replayed

It seems like something we need to ensure in our fork of Helios

Reply:
> (rather than asking for the slot from the verifier contract)

Do you think we should skip the `verifyProof` part and only call `IHelios(helios).getStorageSlot(_head, publicValues.contractAddress, publicValues.slotKey)`?